### PR TITLE
Implemented writeYaml step

### PR DESCRIPTION
**Pull request by witokondoria**
Delegates to snakeyaml the task of serializing any Object to a yaml file

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -28,7 +28,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>1.609.1</version>
+        <version>2.19</version>
```
Comment (rsandell) [inline pom.xml]: Why do you need this?
I don't see any reason for it. As I've said in previous PRs; When we bump anything here I want to see a more rigorous cleanup for example moving to more recent workflow.version and cleaning up all `@Inject` etc. And that should go in a separate PR.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		}
+
+		@Override
+		public String getDisplayName() {
+			return "Write a yaml from an object.";
+		}
+
+		@Override
+		public Step newInstance(Map<String, Object> arguments) throws Exception {
+			WriteYamlStep step = new WriteYamlStep();
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: I doubt this is needed for this step, skip it and just put `file` and `data` as required params in the `@DataBoundConstructor`.
The special handling here that you've seen in other steps are because they have an _either or_ relation that can't be expressed in a better way.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+		@StepContextParameter
+		private transient FilePath ws;
+
+		@Inject
+		private transient WriteYamlStep step;
+
+		@Override
+		protected Void run () throws Exception {
+            FilePath path = null;
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Bad formating.

File: src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepTest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Test
+    public void writeInvalidMap() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+                "def l = [] \n" +
+                        "node('slaves') {\n" +
+                        "  writeYaml file: 'test', data: /['a': ]/ \n" +
+                        " sh 'cat test' \n" +
```
Comment (rsandell) [inline src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepTest.java]: There is a feature coming soon to pipeline which stores the parameters of steps, that might risk printing the step to the log, which means your assert might not assert on the thing it think it does.

It would be more precise if you archived the file and read it back in in the test to verify the contents. But it's just a _nit_.

File: pom.xml (hunk 1 above)
Comment (witokondoria) [inline pom.xml]: The real reason was to bump surefire plugin up to 2.19.1 so testing could be done with -Dtest=method1#method2. Will revert the change

File: src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepTest.java (hunk 4 above)
Comment (witokondoria) [inline src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStepTest.java]: Will readYaml in every test

**Comment by witokondoria** [comment]
Requested changes were adressed. @rsandell, could you review them?

**Comment by witokondoria** [comment]
Completely understandable, better block a bug than fix it at a later stage.
I believe its easier/leaner to limit the classes to dump (ala script-approval). Inspecting the object to check if every getter is whitelisted does not seem really clear.

Have added a recursive (:imp:) isValidObjectType method to check wether the object to dump is "safe". WDYT?

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            (obj instanceof Number) || (obj instanceof String) ||
+            (obj instanceof URL) || (obj instanceof Calendar) ||
+            (obj instanceof Date) || (obj instanceof UUID) ||
+            (obj == null)) {
+            return true;
+        } else if (obj instanceof Map)  {
+            Boolean ret = true;
+            Set entry = ((Map) obj).entrySet();
+            Iterator iterator = entry.iterator();
+            while (iterator.hasNext() && ret) {
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: This looks a bit convoluted, something like this would be more readable and simpler I think

```java
for (Map.Entry entry : map.entrySet()) {
    if (!isValidObject(entry.getKey()) || !isValidObject(entry.getValue())) {
        return false;
    }
}
```

It also seems like you are overwriting the `ret` var every time in the loop, so effectively you are only checking the last entry in the map which my suggestion also solves.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                Object next = iterator.next();Object es = ((Map) obj).get(next);
+                ret = (isValidObjectType(((Map.Entry)next).getKey()) && isValidObjectType(((Map.Entry)next).getValue()));
+            }
+            return ret;
+        } else if (obj instanceof Collection)  {
+            Boolean ret = true;
+            Iterator iterator = ((Collection) obj).iterator();
+            while (iterator.hasNext() && ret) {
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Same simplification and issue here as for the Map.
Suggestion:
```java
for (Object o : ((Collection)obj)) {
    if (!isValidObject(o)) {
        return false;
    }
}
```

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/config.properties [hunk 7]
```diff
@@ -0,0 +1,4 @@
+file.title=File
+file.description=File to save as YAML
+text.title=Data
+text.description=Object to serialize yo YAML
```
Comment (daniel-beck) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/config.properties]: Yo dawg, I heard yo and yo dawg like yo yos.

File: pom.xml [hunk 8]
```diff
@@ -175,4 +175,11 @@
         <url>https://github.com/jenkinsci/pipeline-utility-steps-plugin</url>
         <tag>HEAD</tag>
     </scm>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>repo.jenkins-ci.org</id>
+            <url>http://repo.jenkins-ci.org/public/</url>
```
Comment (oleg-nenashev) [inline pom.xml]: https 🐋

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+import java.nio.file.FileAlreadyExistsException;
+import java.util.*;
+
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Writes a yaml file from the workspace.
+ *
+ * @author Javier DELGADO &lt;witokondoria@gmail.com&gt;.
+ */
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: `since TODO` would be useful since we make attempts to visualize extension javadocs

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+public class WriteYamlStep extends AbstractStepImpl {
+
+    private String file;
+    private Object data;
+
+    @DataBoundConstructor
+    public WriteYamlStep(@Nonnull String file, @Nonnull Object data) {
+        if (file == null) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: StringUtils.isBlank()

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            throw new IllegalArgumentException("file parameter must be provided to writeYaml");
+        }
+        this.file = file;
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: truncate spaces?

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Does it also make sense to verify the patch and to ensure nobody passes an absolute one or a relative with traversal to upper levels?

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    public String getFile() {
+        return file;
+    }
+
+    /**
+     * Name of the yaml file to write.
+     *
+     * @param file file name
+     */
+    public void setFile(String file) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Do you need setters? If no, I would advice to either remove them and make the fields final OR to convert the class to `@DataBoundSetter`s

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        this.data = data;
+    }
+
+    private boolean isValidObjectType(Object obj) {
+        if ((obj instanceof Boolean) || (obj instanceof Character) ||
+            (obj instanceof Number) || (obj instanceof String) ||
+            (obj instanceof URL) || (obj instanceof Calendar) ||
+            (obj instanceof Date) || (obj instanceof UUID) ||
+            (obj == null)) {
+            return true;
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: It is strange, error-prone (E.g. GString, Long, etc.) and ineffective. Why would you need it?

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 5 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: True that, Its resoult of a late commit. Will fix it.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 6 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: ditto

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 11 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: want to trim the string? Forcing a path from within the current current workspace seems fine

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+     * @param data to parse
+     */
+    public void setData(Object data) {
+        this.data = data;
+    }
+
+    private boolean isValidObjectType(Object obj) {
+        if ((obj instanceof Boolean) || (obj instanceof Character) ||
+            (obj instanceof Number) || (obj instanceof String) ||
+            (obj instanceof URL) || (obj instanceof Calendar) ||
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: I had some concerns about this, but it seems it's safe enough:
https://bitbucket.org/asomov/snakeyaml/src/6c6516c003eb4902a0c4c308100c26b99e2ef99c/src/main/java/org/yaml/snakeyaml/representer/SafeRepresenter.java?at=default&fileviewer=file-view-default#SafeRepresenter.java-331

File: pom.xml [hunk 15]
```diff
@@ -175,4 +175,11 @@
         <url>https://github.com/jenkinsci/pipeline-utility-steps-plugin</url>
         <tag>HEAD</tag>
     </scm>
+
+    <pluginRepositories>
```
Comment (daniel-beck) [inline pom.xml]: Time for a new parent pom before trying to fix this.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 13 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Its not a complete and final list, but a one less strict than plain Maps, with objects [already whitelisted](https://github.com/jenkinsci/script-security-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist) when toJson invocations:
Long is covered by Number
GStrings hasnt been even tested :(

Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: Its not a complete and final list, but a one less strict than plain Maps, with objects [already whitelisted](https://github.com/jenkinsci/script-security-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist) when toJson invocations:
Long is covered by Number
GStrings hasnt been even tested :(

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/config.properties (hunk 7 above)
Comment (witokondoria) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep/config.properties]: LOL

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 13 above)
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: My guess is that the `GString`s will be evaluated/expanded in CPS/sandbox before entering the java context, but it needs to be tested non the less.

something like:

```groovy
writeYaml data: "${currentBuild.rawBuild}", file: 'data'
```
might be enough

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 9 above)
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: not needed. it's not present in any other of the steps.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 13 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: will add tests tomorrow!

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java (hunk 11 above)
Comment (witokondoria) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java]: From the [writeFile Step](https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/steps/WriteFileStep.java#L100), paths are not verified (so traversal is also possible).

Would you prefer a higer level implementation for this check?

**Comment by witokondoria** [comment]
Most of the changes were adressed (path check is unimplemented). Im willing to squash the commits in one, but have left all of them to easen reviews.
